k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageCryptoPresenter.java

```diff
@@ -0,0 +1,226 @@
+            }
+        }
+
+        return true;
+    }
+
+    private void showMessageCryptoWarning(MessageTopView messageView, Account account,
+            MessageViewInfo messageViewInfo, @StringRes int warningStringRes) {
+        if (overrideCryptoWarning) {
+            messageView.showMessage(account, messageViewInfo);
+            return;
+        }
+        Drawable providerIcon = getOpenPgpApiProviderIcon(messageView.getContext(), account);
+        messageView.showMessageCryptoWarning(messageViewInfo, providerIcon, warningStringRes);
+    }
+
+
+    @Override
+    public void onCryptoClick() {
+        if (cryptoResultAnnotation == null) {
+            return;
+        }
+        MessageCryptoDisplayStatus displayStatus =
+                MessageCryptoDisplayStatus.fromResultAnnotation(cryptoResultAnnotation);
+        switch (displayStatus) {
+            case LOADING:
+                // no need to do anything, there is a progress bar...
+                break;
+            case UNENCRYPTED_SIGN_UNKNOWN:
+                launchPendingIntent(cryptoResultAnnotation);
+                break;
+            default:
+                displayCryptoInfoDialog(displayStatus);
+                break;
+        }
+    }
+
+    @SuppressWarnings("UnusedParameters") // for consistency with Activity.onActivityResult
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode != REQUEST_CODE_UNKNOWN_KEY) {
+            throw new IllegalStateException("got an activity result that wasn't meant for us. this is a bug!");
+        }
+
+        if (resultCode != Activity.RESULT_OK) {
+            return;
+        }
+
+        messageCryptoMvpView.restartMessageCryptoProcessing();
+    }
+
+    private void displayCryptoInfoDialog(MessageCryptoDisplayStatus displayStatus) {
+        messageCryptoMvpView.showCryptoInfoDialog(displayStatus);
+    }
+
+    private void launchPendingIntent(CryptoResultAnnotation cryptoResultAnnotation) {
+        try {
+            PendingIntent pendingIntent = cryptoResultAnnotation.getOpenPgpPendingIntent();
+            if (pendingIntent != null) {
+                messageCryptoMvpView.startPendingIntentForCryptoPresenter(
+                        pendingIntent.getIntentSender(), REQUEST_CODE_UNKNOWN_KEY, null, 0, 0, 0);
+            }
+        } catch (IntentSender.SendIntentException e) {
+            Log.e(K9.LOG_TAG, "SendIntentException", e);
+        }
+    }
+
+    public void onClickShowCryptoKey() {
+        try {
+            PendingIntent pendingIntent = cryptoResultAnnotation.getOpenPgpPendingIntent();
+            if (pendingIntent != null) {
+                messageCryptoMvpView.startPendingIntentForCryptoPresenter(
+                        pendingIntent.getIntentSender(), null, null, 0, 0, 0);
+            }
+        } catch (IntentSender.SendIntentException e) {
+            Log.e(K9.LOG_TAG, "SendIntentException", e);
```

Comment:
Nothing. I guess we could show a toast, but that's going to be one of them unhelpful "internal error" ones...